mountpoint-s3/src/prefetch/task.rs

```diff
@@ -14,36 +15,62 @@ pub struct RequestTask<E: std::error::Error> {
     start_offset: u64,
     total_size: usize,
     part_queue: PartQueue<E>,
+    read_window_updater: Sender<usize>,
 }
 
 impl<E: std::error::Error + Send + Sync> RequestTask<E> {
-    pub fn from_handle(task_handle: RemoteHandle<()>, size: usize, offset: u64, part_queue: PartQueue<E>) -> Self {
+    pub fn from_handle(
+        task_handle: RemoteHandle<()>,
+        size: usize,
+        offset: u64,
+        part_queue: PartQueue<E>,
+        read_window_updater: Sender<usize>,
+    ) -> Self {
         Self {
             task_handle: Some(task_handle),
             remaining: size,
             start_offset: offset,
             total_size: size,
             part_queue,
+            read_window_updater,
         }
     }
 
-    pub fn from_parts(parts: impl IntoIterator<Item = Part>, offset: u64) -> Self {
+    pub fn from_parts(parts: impl IntoIterator<Item = Part>, offset: u64, read_window_updater: Sender<usize>) -> Self {
         let mut size = 0;
         let (part_queue, part_queue_producer) = unbounded_part_queue();
         for part in parts {
             size += part.len();
-            part_queue_producer.push(Ok(part));
+            part_queue_producer.push(Ok(part), None);
         }
         Self {
             task_handle: None,
             remaining: size,
             start_offset: offset,
             total_size: size,
             part_queue,
+            read_window_updater,
         }
     }
 
+    // Push a given list of parts in front of the part queue
+    pub async fn push_front(&mut self, mut parts: Vec<Part>) -> Result<(), PrefetchReadError<E>> {
+        // Merge all parts into one single part.
```

Comment:
This could be expensive (especially because of checksumming). Couldn't we just keep a vector of parts?

Reply:
My thinking is that it's fairly cheap compare to creating a new request and this way we don't have to change the part queue structure. I will see if we can improve this.